### DISCUSSION TITLE
Timestamps in data

### DISCUSSION THREAD

**Discussion by roaldarbol** [talmolab/sleap#1885] [General]
Hi again! :-)
I've tried to look through the documentation and the `.slp` file itself, but I can't find any timestamps. Does SLEAP not keep the timestamps from the video?

**Comment by roomrys** [comment]
Hi @roaldarbol,

SLEAP does not keep the timestamps. If the video you loaded in SLEAP has one of these extensions:
https://github.com/talmolab/sleap/blob/28c34e22e0cf78e1774476d0ac76c7ea0b4814fe/sleap/io/video.py#L354
then the `Video` object it was loaded into has a `fps` property
https://github.com/talmolab/sleap/blob/28c34e22e0cf78e1774476d0ac76c7ea0b4814fe/sleap/io/video.py#L432-L435
that returns the frames per second which you can use in conjunction with the frame index to calculate a relative timestamp (video starts at time 0).

```python
(sleap) roomrys:~$ipython

In [1]: from sleap import Labels

In [2]: ds = "/Users/roomrys/Projects/sleap-datasets/drosophila-melanogaster-
   ...: courtship/courtship_labels copy.slp"

In [3]: labels = Labels.load_file(ds)

In [4]: labeled_frame = labels.labeled_frames[0]

In [5]: video = labeled_frame.video

In [6]: frame_idx = labeled_frame.frame_idx

In [7]: video.fps
Out[7]: 25.0

In [8]: frame_idx / video.fps
Out[8]: 0.4
```

Sorry about that,
Liezl

**Reply by roaldarbol** [reply]
Hi @roomrys! 
Ah, I see. Yeah, I guess the video files don't encode the timestamps (which is a pity!). Since most cameras have a rolling shutter, the `fps` approach is quite a rough estimate (as `fps` fluctuate a fair amount throughout); I actually ended up outputting time stamps with Bonsai whilst recording to have the precise ones. But I completely see how it's out of the SLEAP remit, I'de just forgotten all about that. ;-)

**Comment by roaldarbol** [comment]
Just came across this which can be achieved with `ffprobe`: https://superuser.com/a/1418998.

**Reply by roaldarbol** [reply]
Looked into it a bit more, and it doesn't work - it simply uses the `fps` to "re-create" the playback timestamp, not the actual time of recording.

**Comment by talmo** [comment]
Hi @roaldarbol,

In case you haven't found your way over to this issue yet, there's a good discussion over here on the topic of video timestamps/accurate+fast seeking/encoding and etc: https://github.com/janclemenslab/napari-video/issues/3

Hope that helps!

FWIW, our setup is to store the timestamps separately from Bonsai as well. We pretty much entirely disregard the video FPS and "timestamps" stored in the video container metadata.

Talmo

**Reply by roaldarbol** [reply]
Hi @talmo!

I indeed hadn't! Super interesting, and I learned a lot there! 

The more I've immersed myself in all this video stuff, the more I feel like it would be great to bring together a team to (get funding for and) develop a super robust video acquisition software that can handle a bunch of different camera types, simultaneous recording from multiple cameras; as you mentioned, a lot of it comes down to lots of poor acquisition software, I've never had access to high-end cameras or proprietary video acquisition software. Bonsai's great, but the fact that it's only cross-platform. with a lot of hassle has turned me away from it in the past. Something simple, that works for everyone, video only.

Cheers, Mikkel

**Reply by talmo** [reply]
You might like: https://github.com/ksseverson57/campy

**Reply by roaldarbol** [reply]
Ideally I'd like to see something that works with webcams too, GoPros, Raspberry Pi cams, etc. (as well as industrial cameras). It's simply much more common in smaller, poorly funded labs. Something that could make it just as easy to gain the advantages on a laptop in the field. :-)

**Reply by roaldarbol** [reply]
I've brought up the idea to the FreeMoCap folks, as FreeMoCap also features a recording interface, and that work might be well worth building on.
